rooms.js

```diff
@@ -163,6 +165,107 @@ var Room = (function () {
 		}
 		return true;
 	};
+	//mute handling
+	Room.prototype.runMuteTimer = function () {
+		if (this.muteTimer || this.mutedUsers.length === 0) return;
+		var timeUntilExpire = this.mutedUsers[0].time - Date.now();
+		if (timeUntilExpire <= 0) {
+			var user = Users(this.mutedUsers[0].userid);
+			if (user) {
+				user.popup("Your mute has expired.");
+				this.unmute(user);
+			} else {
+				this.mutedUsers.splice(0, 1);
+				this.runMuteTimer();
+			}
+		} else {
+			var self = this;
+			this.muteTimer = setTimeout(function () {
+				var user = Users(self.mutedUsers[0].userid);
+				if (user) {
+					user.popup("Your mute has expired.");
+					self.unmute(user);
+				} else {
+					self.mutedUsers.splice(0, 1);
+					self.runMuteTimer();
+				}
+			}, timeUntilExpire);
+		}
+	};
+	Room.prototype.isMuted = function (user) {
+		if (!user) return;
+		if (this.mutedUsers) {
+			for (var i = 0; i < this.mutedUsers.length; i++) {
+				if (user.userid === this.mutedUsers[i].userid ||
+					user.guestNum === this.mutedUsers[i].guestNum ||
+					(user.autoconfirmed && user.autoconfirmed === this.mutedUsers[i].autoconfirmed)) {
+					return this.mutedUsers[i].userid;
+				}
+			}
+		}
+	};
+	Room.prototype.getMuteTime = function (user) {
+		var userid = this.isMuted(user);
+		if (!userid) return;
+		for (var i = 0; i < this.mutedUsers.length; i++) {
+			if (userid === this.mutedUsers[i].userid) {
+				return this.mutedUsers[i].time - Date.now();
+			}
+		}
+	};
+	Room.prototype.mute = function (user, setTime) {
+		var userid = user.userid;
+
+		if (!setTime) setTime = 7 * 60000; // default time: 7 minutes
+		if (setTime < 1) setTime = 1; // mostly to prevent bugs
+		if (setTime > 90 * 60000) setTime = 90 * 60000; // limit 90 minutes
+
+		// If the user is already muted, the existing queue position for them should be removed
+		if (this.isMuted(user)) this.unmute(user);
+
+		// Place the user in a queue for the unmute timer
+		for (var queue = 0; queue < this.mutedUsers.length + 1; queue++) {
+			var timeRemaining = this.mutedUsers[queue] ? this.mutedUsers[queue].time - Date.now() : Infinity;
+			if (setTime < timeRemaining) {
+				var buffer = Object.create(null);
```

Comment:
You should be doing something more like:

	var buffer = {
		userid: userid,
		time: Date.now() + setTime,
		guestNum: user.guestNum,
		autoconfirmed: user.autoconfirmed
	};

Although I'd recommend calling it `entry` rather than `buffer`.